sympy/physics/mechanics/joint.py

```diff
@@ -0,0 +1,257 @@
+
+        self._coordinates = self._generate_coordinates(coordinates)
+        self._speeds = self._generate_speeds(speeds)
+        self._kdes = self._generate_kdes()
+
+        self._parent_axis = self._axis(parent, parent_axis)
+        self._child_axis = self._axis(child, child_axis)
+
+        self._parent_joint = self._locate_joint_pos(parent,parent_joint_pos)
+        self._child_joint = self._locate_joint_pos(child, child_joint_pos)
+
+        self._orient_frames()
+        self._set_angular_velocity()
+        self._set_linear_velocity()
+
+    def __str__(self):
+        return self._name
+
+    def __repr__(self):
+        return self.__str__()
+
+    def parent(self):
+        """ Parent body of Joint."""
+        return self._parent
+
+    def child(self):
+        """ Child body of Joint."""
+        return self._child
+
+    def coordinates(self):
+        """ List of coordinates of Joint."""
+        return self._coordinates
+
+    def speeds(self):
+        """ List of speeds of Joint."""
+        return self._speeds
+
+    def kdes(self):
+        """ KDE of Joint."""
+        return self._kdes
+
+    @abstractmethod
+    def _generate_coordinates(self, coordinates):
+        """ Generate list of coordinates."""
+        pass
+
+    @abstractmethod
+    def _generate_speeds(self, speeds):
+        """ Generate list of speeds. """
+        pass
+
+    @abstractmethod
+    def _orient_frames(self):
+        """Orient frames as per the joint"""
+        pass
+
+    @abstractmethod
+    def _set_angular_velocity(self):
+        pass
+
+    @abstractmethod
+    def _set_linear_velocity(self):
+        pass
+
+    def _generate_kdes(self):
+        kdes = []
+        t = dynamicsymbols._t
+        for i in range(len(self._coordinates)):
+            kdes.append(-self._coordinates[i].diff(t) + self._speeds[i])
+        return kdes
+
+    def _axis(self, body, ax):
+        if ax is None:
+            ax = body.frame.x
+            return ax
+        if not isinstance(ax, Vector):
+            raise TypeError("Axis must be of type Vector. Example-> A.x wehere 'A' is ReferenceFrame.")
+        return ax
+
+    def _locate_joint_pos(self, body, joint_pos):
+        if joint_pos is None:
+            joint_pos = Vector(0)
+        if not isinstance(joint_pos, Vector):
+            raise ValueError('Joint Position must be supplied as Vector.')
+        return body.masscenter.locatenew(self._name + '_' + body.name + '_joint', joint_pos)
+
+    def _set_masscenter_pos(self, dist):
+        self._child.masscenter.set_pos(self._parent.masscenter, dist)
+
+    def _align_axis(self):
+        axis1 = self._parent_axis
+        axis2 = self._child_axis
+        angle = acos(dot(axis1, axis2)/(axis1.magnitude() * axis2.magnitude()))
+        if angle!=0:
+            I = ReferenceFrame('I')
+            I.orient_axis(self._parent.frame, axis1, angle)
+            self._child.frame.orient_axis(I, axis1, self._coordinates[0])
```

Comment:
@moorepants Please check if this implementation is correct, I have also added a very simple test case(test_pinjoint_arbitrary_axis) to check if axis are being aligned,.

Reply:
You should check if it is correct. You'll need to work out the kinematics manually and make sure they match.